contracts/PooledStaking.sol

```diff
@@ -0,0 +1,522 @@
+      }
+
+      // get next available id
+      uint id = deallocationCount;
+      deallocationCount++;
+
+      uint next = current.next;
+
+      // point to our new deallocation
+      current.next = id;
+
+      // insert next item after this one
+      insertAfter = id;
+
+      deallocations[id] = Deallocation(
+        requested, deallocateAt, contractAddress, msg.sender, next
+      );
+
+      // increase pending deallocation amount so we keep track of final allocation
+      uint newPending = staker.pendingDeallocations[contractAddress].add(requested);
+      staker.pendingDeallocations[contractAddress] = newPending;
+    }
+  }
+
+  function pushBurn(address contractAddress, uint amount) external onlyInternal {
+
+    // add new burn
+    burns[burnCount] = Burn(amount, now, contractAddress, 0);
+
+    // TODO: recheck this
+    // burnCount and firstBurn can be equal only when they're both 0
+    // i.e. when pushing the first burn ever
+    // if those are different, burnCount - 1 will be >= 0 (will not underflow)
+    // but we should check that it doesn't point to an empty slot (processed & deleted burn)
+    if (burnCount != firstBurn && burns[burnCount - 1].burnedAt > 0) {
+      // set previous burn's next to current burn id
+      burns[burnCount - 1].next = burnCount;
+    } else {// otherwise this is the only existing burn, therefore it is the first one as well
+      firstBurn = burnCount;
+    }
+
+    // update counter
+    ++burnCount;
+  }
+
+  function pushReward(address contractAddress, uint amount, address from) external onlyInternal {
+
+    // transfer tokens from specified contract to us
+    // token transfer should be approved by the specified address
+    Vault.deposit(token, from, amount);
+
+    // add new reward
+    rewards[rewardCount] = Reward(amount, now, contractAddress, 0);
+
+    // TODO: recheck this
+    // similar to pushBurn
+    if (rewardCount != firstReward && rewards[rewardCount - 1].rewardedAt > 0) {
+      // set previous reward's next pointer to current id
+      rewards[rewardCount - 1].next = rewardCount;
+    } else {
+      firstReward = rewardCount;
+    }
+
+    // update counter
+    ++rewardCount;
+  }
+
+  function processPendingActions() public returns (bool) {
+
+    while (true) {
+
+      Burn storage burn = burns[firstBurn];
+      Deallocation storage deallocation = deallocations[firstDeallocation];
+      Reward storage reward = rewards[firstReward];
+
+      bool canBurn = burn.burnedAt != 0;
+      bool canDeallocate = deallocation.deallocateAt != 0 && deallocation.deallocateAt <= now;
+      bool canReward = reward.rewardedAt != 0;
+
+      if (!canBurn && !canDeallocate && !canReward) {
+        // everything is processed
+        break;
+      }
+
+      if (
+        canBurn &&
+        (!canDeallocate || burn.burnedAt < deallocation.deallocateAt) &&
+        (!canReward || burn.burnedAt < reward.rewardedAt)
+      ) {
+        // O(n*m)
+        _processFirstBurn();
```

Comment:
i'd guess we need to go 1 level deeper with gas estimations since at this level only n can be evaluated contracts.length